From d5349a0505724bcf078c19d2a49a8b738e13b894 Mon Sep 17 00:00:00 2001
From: Sailaja <sbandarupalli@habana.ai>
Date: Mon, 20 Apr 2020 14:08:42 +0300
Subject: [PATCH] [SW-11867] goya2: preboot: pll config changes

Added goya2 specific pll changes for arc boot flows

In goya2, pll coefficients and programming are changed
created goya2 specific functions which overrides goya1/gaudi1 functionality

Change-Id: Ib8d8bebdb516dabeac2767200f60e46fd07eed8e
Signed-off-by: Sailaja <sbandarupalli@habana.ai>
---

diff --git a/Makefile b/Makefile
index 3b9b8f8..c52af14 100644
--- a/Makefile
+++ b/Makefile
@@ -30,6 +30,7 @@
 
 PREBOOT_SRCS_C := $(shell ls common/*.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/common.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/pll.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/$(HL_TARGET).c)
 PREBOOT_APP_S := $(shell ls app/crt0_preboot.S)
 PREBOOT_APP_C := $(shell ls app/preboot.c)
diff --git a/common/pll.c b/common/pll.c
index 91d37c2..9891c4b 100644
--- a/common/pll.c
+++ b/common/pll.c
@@ -18,10 +18,7 @@
 #define HL_PLL_LOCK_INTR	0x128
 #define HL_PLL_DATA_CHNG	0x130
 #define HL_PLL_RST		0x134
-#define HL_PLL_DIV_FACTOR	0x200
 #define HL_PLL_DIV_FACTOR_CMD	0x220
-#define HL_PLL_DIV_SEL		0x280
-#define HL_PLL_DIV_EN		0x2a0
 #define HL_PLL_CLK_GATER	0x300
 
 #define MIN_CYCLES_PER_ITER	5
@@ -41,7 +38,7 @@
 		;
 }
 
-u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
+__attribute__((weak)) u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
 {
 	u32 nr, nf, od, div_sel, factor, offset;
 	u64 out_freq = SYS_REF_CLOCK;
diff --git a/common/pll.h b/common/pll.h
index e1306ee..bf69361 100644
--- a/common/pll.h
+++ b/common/pll.h
@@ -8,6 +8,9 @@
 #ifndef _HL_PLL_H_
 #define _HL_PLL_H_
 
+#define HL_PLL_DIV_FACTOR	0x200
+#define HL_PLL_DIV_SEL		0x280
+#define HL_PLL_DIV_EN		0x2a0
 #define SYS_REF_CLOCK		50000000ULL
 #define USEC_PER_SEC		1000000
 #define BYPASS_CYCLES_PER_USEC	(SYS_REF_CLOCK / USEC_PER_SEC)
diff --git a/goya2_arc_gnu/common.h b/goya2_arc_gnu/common.h
index aa3a2b3..b7c08a0 100644
--- a/goya2_arc_gnu/common.h
+++ b/goya2_arc_gnu/common.h
@@ -25,3 +25,10 @@
 #define mmTPC_PLL_BASE		0xE01000
 
 #define FUSE_WORDS_PER_BANK	0x40
+
+/*
+ * extracts "num" number of bits from position "pos" in the
+ * given number "value"
+ */
+#define EXTRACT_BITS(value, pos, num) ((value & (((1 << num) - 1) << pos))\
+						>> pos)
diff --git a/goya2_arc_gnu/pll.c b/goya2_arc_gnu/pll.c
new file mode 100644
index 0000000..ea7f16a
--- /dev/null
+++ b/goya2_arc_gnu/pll.c
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2016-2020 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include <stddef.h>
+#include <preboot.h>
+#include <pll.h>
+#include "common.h"
+
+
+#define GOYA2_PLL_CFG_DIV		0x104
+
+u32 hl_pll_get_freq(u32 pll_base, enum hl_pll_div_out div)
+{
+	u32 div_sel, factor, offset;
+	u32 reg, fbdiv, refdiv, postdiv1, postdiv2;
+	u64 out_freq = SYS_REF_CLOCK;
+
+	offset = div * sizeof(u32);
+	div_sel = regmap_readl(pll_base + HL_PLL_DIV_SEL + offset);
+
+	if (div_sel == HL_PLL_PLL_CLK || div_sel == HL_PLL_DIV_PLL) {
+		reg = regmap_readl(pll_base + GOYA2_PLL_CFG_DIV);
+		/* fbdiv : bits 0-11 */
+		fbdiv = EXTRACT_BITS(reg, 0, 12);
+		/* postdiv1 : bits 16-18 */
+		postdiv1 = EXTRACT_BITS(reg, 16, 3);
+		/* postdiv2 : bits 20-22 */
+		postdiv2 = EXTRACT_BITS(reg, 20, 3);
+		/* refdiv : bits 24-29 */
+		refdiv  = EXTRACT_BITS(reg, 24, 6);
+		out_freq = SYS_REF_CLOCK * fbdiv;
+		out_freq /= refdiv * postdiv1 * postdiv2;
+	}
+
+	if (div_sel == HL_PLL_DIV_REF || div_sel == HL_PLL_DIV_PLL) {
+		factor = regmap_readl(pll_base + HL_PLL_DIV_FACTOR + offset);
+		out_freq /= factor + 1;
+	}
+
+	/*
+	 * TODO : The clock relaxation is not accounted here.
+	 * We might need to deal with it in the future.
+	 */
+
+	return out_freq;
+}
+
